File: frontend/src/api/categories.ts
import { api } from "./client";

export interface CategoryOut {
  id: number;
  name: string;
  item_count: number;
}

export function listCategories(): Promise<CategoryOut[]> {
  return api.get<CategoryOut[]>("/api/categories");
}

export function createCategory(name: string): Promise<CategoryOut> {
  return api.post<CategoryOut>("/api/categories", { name });
}

export function renameCategory(id: number, name: string): Promise<CategoryOut> {
  return api.patch<CategoryOut>(`/api/categories/${id}`, { name });
}

export function deleteCategory(id: number): Promise<void> {
  return api.del<void>(`/api/categories/${id}`);
}

File: backend/app/routers/categories.py
"""Category endpoints."""

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from app.db import get_db
from app.deps import get_current_user
from app.models import User
from app.schemas import CategoryCreate, CategoryOut, CategoryUpdate

router = APIRouter(prefix="/api/categories", tags=["categories"])


@router.get("", response_model=list[CategoryOut])
def list_categories(
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> list[CategoryOut]:
    raise HTTPException(status_code=501, detail="categories #14 implements this")


@router.post("", response_model=CategoryOut, status_code=201)
def create_category(
    payload: CategoryCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> CategoryOut:
    raise HTTPException(status_code=501, detail="categories #14 implements this")


@router.patch("/{id}", response_model=CategoryOut)
def update_category(
    id: int,
    payload: CategoryUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> CategoryOut:
    raise HTTPException(status_code=501, detail="categories #14 implements this")


@router.delete("/{id}", status_code=204)
def delete_category(
    id: int,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> None:
    raise HTTPException(status_code=501, detail="categories #14 implements this")
